feat: Modal Component

callinmullaney opened the pull request:
**Summary**
- Adds a boilerplate modal component to compound.

This PR fixes/implements the following **bugs/features**
- Implements basic principles of collapsible sections (https://inclusive-components.design/collapsible-sections/) to create a modal component that is accessible and keyboard-friendly.

Explain the **motivation** for making this change. What existing problem does the pull request solve?
- Modals are a very common method of displaying content. It has been a prioritized component defined within the Emulsify UI-Kit initiative. 

**Documentation Update (required)**
- None

**How to review this PR**
- [ ] checkout branch and run `npm install` then `npm run storybook`
- [ ] Go to `Molecules/Modal` and confirm you can see multiple modal options. The functionality for each should be the same. The differences center around the markup that triggers the modal which can be any type of markup. It does not have to be triggered from a button/link only.
- [ ] Use your keyboard to focus and trigger the modal. Verify the link within the modal is focus-able once the modal is active and pressing the `esc` closes the modal
- [ ] Review the twig and styles and verify they are a good starting point for future projects and arent too opinionated.

mndonx commented on components/02-molecules/modal/_modal.scss:
```diff
@@ -0,0 +1,95 @@
+// Base modal styles. Additional classes accomodate for custom app implementations.
```
*accommodate

mndonx commented on components/02-molecules/modal/_modal.scss:
```diff
@@ -0,0 +1,95 @@
+  transition: opacity 300ms;
+  padding: $space;
+  visibility: hidden;
+  z-index: 1000;
```
I order styles alphabetically, but I haven't checked to see what the latest is on that -- I've seen reference to "position styles first." Do you have a preference?

callinmullaney replied:
I'm not actually sure what I prefer but agree with your comment here that there should be some sort of ordering. I updated the properties to be alphabetical.

mndonx commented on components/02-molecules/modal/modal.stories.js:
```diff
@@ -0,0 +1,38 @@
+export const modalMedia = () => {
+  return modalTwig({
+    ...modalData,
+    modal__toggle: '<img src="https://placeimg.com/320/180/any">',
```
I know it's just an example, but maybe an empty alt attribute is still worth throwing in here?

mndonx commented on components/02-molecules/modal/modal.twig:
```diff
@@ -0,0 +1,84 @@
+{% set modal__modifiers = modal__modifiers|default([]) %}
+
+{% if modal__id is empty %}
+	{% set modal__id = random(100000) %}
```
extra space here?

mndonx reviewed (changes requested):
This looks great! I have some questions inline, and I also have  a question about adherence to accessibility in this stripped-down format. I know we want to keep it simple, but I wonder if, bare minimum, it still passes contrast test. See attached. 

Also - does it seem like the close button isn't appearing?
<img width="547" alt="Screen Shot 2022-05-08 at 12 12 54 AM" src="https://user-images.githubusercontent.com/2343957/167282852-b47e2f1c-fecc-49a0-b4c6-0730963a8e0b.png">

callinmullaney commented:
@mndonx Thank you for the thorough review! I addressed your inline comments and added variable/block documentation for the twig file. I also made adjustments to the default colors we're using to help address the contrast warnings you mentioned. This will affect all components so hopefully should eliminate this accessibility error for others once merged.

One accessability error you will come across when re-testing is `Ensures tabindex attribute values are not greater than 0` - This is intentional as it plays into the keyboard-trap JS and allows the modal to draw focus above everything else on a page.

Lastly, I fixed the icon display issue. Looks like the webpack script needed to explicitly point to the `./images/icons` folder and not just `./images`

callinmullaney commented:
@joetower @ryanhagerty Adding you to this PR to help move things along if you have time. Can you confirm I addressed Amanda's feedback from above and let me know if this is ready for an accessibility review?

ryanhagerty commented on components/02-molecules/modal/_modal.scss:
```diff
@@ -0,0 +1,95 @@
+  position: fixed;
+  top: 0;
+
+  &--active {
```
Probably want to unnest this to `modal__overlay--active` to match updated best practices

ryanhagerty commented on components/02-molecules/modal/_modal.scss:
```diff
@@ -0,0 +1,95 @@
+  transition: opacity 300ms;
+  visibility: hidden;
+  width: 100vw;
+  z-index: 1000;
```
Are all the other compound component properties in alphabetical order? I don't think it matters with the upcoming auto-sorter, but just curious if this was a thing.

ryanhagerty commented on components/02-molecules/modal/modal.js:
```diff
@@ -0,0 +1,150 @@
+          response = `<p>Could not process request. Error code: ${request.status}</p>`;
+        }
+      };
+      request.open('GET', url, false);
```
Why are we including ajax functionality for the modal? This seems very strange to me. The modern standard is `fetch`, some people still prefer `axios`, and even fetching data from a modal seems like an edge case.

We should remove this. Kudos for trying to cover all the bases, but let's let the small amount of people doing this decide how they want to implement fetching data.

ryanhagerty commented on components/02-molecules/modal/modal.js:
```diff
@@ -0,0 +1,150 @@
+              toggleChild.getAttribute('href') !== null
+            ) {
+              e.preventDefault();
+              const request = new XMLHttpRequest();
```
We should take all this out

ryanhagerty commented on components/02-molecules/modal/modal.twig:
```diff
@@ -0,0 +1,107 @@
+    <div {{ bem('content', modal__modifiers, modal__base_class) }}>
+      {% block modal__content %}
+        {% include "@atoms/text/text/01-paragraph.twig" with {
+          paragraph_content: modal_content,
```
Why do we include modal content within a `paragraph` component? How do we know it will be text content? I think this should be a generic `div`

ryanhagerty commented on components/02-molecules/modal/modal.twig:
```diff
@@ -0,0 +1,107 @@
+{% set modal__modifiers = modal__modifiers|default([]) %}
+
+{% if modal__id is empty %}
+	{% set modal__id = random(100000) %}
```
I get why you made it random here - in case there's multiple modals right? But part of the accessibility for having the id is to correspond with the label.

Is a screenreader reading `modal-35827-heading` ideal? This might need a revisit

mndonx replied:
Maybe you could set the random id up top and use it for the id and for the label?

ryanhagerty commented on components/02-molecules/modal/_modal.scss:
```diff
@@ -0,0 +1,95 @@
+.modal__content {
+  @include wrapper($container-max-width: $small);
+
+  > * {
```
I'm a little confused by this. Let me know if I'm getting this wrong: so all the modal content gets piped through the `paragraph` component, which means an example could be:
```
<p>
  <div>content here
    <span>here
      <h2>oh a heading</h2>
    </span>
  </div>
</p>
```
Won't the `p` always be the last child?

ryanhagerty reviewed (changes requested):
Great work Callin - halfway through the review I remembered I helped on this one - it's been a while! lol

Some general comments, but I don't think anything major.
